test/e2e/upgrade_hypershift_operator_test.go

```diff
@@ -0,0 +1,226 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	"github.com/openshift/hypershift/cmd/install"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	"sigs.k8s.io/cluster-api/api/v1beta1"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nodePoolAnnotationCurrentConfig        = "hypershift.openshift.io/nodePoolCurrentConfig"
+	nodePoolAnnotationCurrentConfigVersion = "hypershift.openshift.io/nodePoolCurrentConfigVersion"
+	hostedClusterUpgradeTestLabel          = "hypershift.openshift.io/upgrade-test"
+)
+
+// TestUpgradeHyperShiftOperator validates that a HyperShift Operator upgrade won't unnecessarily cause a node rollout.
+// The test must be triggered via the flag upgrade.run-tests and the HyperShift Operator initial image must be set
+// via the flag e2e.hypershift-operator-initial-image.
+func TestUpgradeHyperShiftOperator(t *testing.T) {
+	if !globalOpts.RunUpgradeTest {
+		// This test should be triggered intentionally by setting the option above,
+		// otherwise skip it.
+		t.SkipNow()
+	} else {
+		var err error
+
+		var mgmtClient crclient.Client
+		var hostedCluster *hyperv1.HostedCluster
+		var hcpNameSpace string
+		var nodePoolsMap map[string]*hyperv1.NodePool
+		var machineDeploymentMap map[string]*v1beta1.MachineDeployment
+
+		hyperShiftOperatorLatestImage := globalOpts.HyperShiftOperatorLatestImage
+		hyperShiftOperatorInitialImage := globalOpts.HyperShiftOperatorInitialImage
+
+		g := gomega.NewWithT(t)
+
+		t.Parallel()
+		ctx, cancel := context.WithCancel(testContext)
+
+		defer cancel()
+
+		cleanup := install.NoOpCleanup
+		t.Cleanup(func() {
+			t.Log("Cleaning up HyperShift Operator")
+			err = cleanup(context.Background())
+			g.Expect(err).ToNot(gomega.HaveOccurred(), "Cleaning up initial HyperShift Operator shouldn't return errors")
+		})
+
+		clusterOpts := globalOpts.DefaultClusterOptions(t)
+
+		t.Log("Starting HyperShift Operator upgrade test")
+		zones := strings.Split(globalOpts.ConfigurableClusterOptions.Zone.String(), ",")
+		if len(zones) >= 3 {
+			t.Log("Sufficient zones available for InfrastructureAvailabilityPolicy HighlyAvailable")
+			clusterOpts.AWSPlatform.Zones = zones
+			clusterOpts.InfrastructureAvailabilityPolicy = string(hyperv1.HighlyAvailable)
+			clusterOpts.NodePoolReplicas = 1
+		}
+
+		clusterOpts.BeforeApply = func(o crclient.Object) {
+			switch obj := o.(type) {
+			case *hyperv1.HostedCluster:
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[hyperv1.ControlPlaneOperatorV2Annotation] = "true"
```

Comment:
I don't think we want to enable CPO V2 for the upgrade test for now.
```suggestion
```

Reply:
fixed